thorlcr/master/thmastermain.cpp

```diff
@@ -358,7 +358,39 @@ class CRegistryServer : public CSimpleInterface
                     EXCLOG(e, "Shutting down slave");
                     e->Release();
                 }
+                if (watchdog)
+                    watchdog->removeSlave(ep);
+            }
+        }
+
+        CTimeMon tm(20000);
+        unsigned numReplied = 0;
+        while (numReplied < slavesRegistered)
+        {
+            unsigned remaining;
+            if (tm.timedout(&remaining))
+            {
+                PROGLOG("Timeout waiting for Shutdown reply from slave(s) (%u replied out of %u total)", numReplied, slavesRegistered);
+                break;
+            }
+            try
+            {
+                rank_t sender;
+                CMessageBuffer msg;
+                if (queryNodeComm().recv(msg, RANK_ALL, shutdownTag, &sender, 50))
+                    numReplied++;
+            }
+            catch (IMP_Exception *e)
+            {
+                // do not log MP link closed exceptions from ending slaves
```

Comment:
Am not sure what kind of IMP_Exception on recv can happen here.

Bit of a problem if there is from a ending slave though, as it will leave numReplied unincremented and loop will never end.